File: WinForms/VB/SpreadsheetMainDemo/DemosRegistration.vb
Imports System.Collections.Generic
Imports System.Windows.Forms
Imports DevExpress.DXperience.Demos

Namespace DevExpress.XtraSpreadsheet.Demos

    Public Class DemosInfo
        Inherits DevExpress.DXperience.Demos.ModulesInfo

        Const MaxMruCount As Integer = 3

        Private Shared ReadOnly DefaultPadding As System.Windows.Forms.Padding = New System.Windows.Forms.Padding(8)

        Private Shared ReadOnly mruList As System.Collections.Generic.List(Of DevExpress.XtraSpreadsheet.Demos.DemoMruItem) = New System.Collections.Generic.List(Of DevExpress.XtraSpreadsheet.Demos.DemoMruItem)()

        Public Shared Sub DoShowModule(ByVal name As String, ByVal groupControl As DevExpress.XtraEditors.GroupControl, ByVal manager As DevExpress.XtraSpreadsheet.Demos.SpreadsheetRibbonMenuManager)
            Dim item As DevExpress.DXperience.Demos.ModuleInfo = DevExpress.XtraSpreadsheet.Demos.DemosInfo.GetItem(name)
            Dim currentCursor As System.Windows.Forms.Cursor = System.Windows.Forms.Cursor.Current
            System.Windows.Forms.Cursor.Current = System.Windows.Forms.Cursors.WaitCursor
            Try
                Dim tutorial As DevExpress.XtraSpreadsheet.Demos.SpreadSheetTutorialControlBase = TryCast(item.TModule, DevExpress.XtraSpreadsheet.Demos.SpreadSheetTutorialControlBase)
                groupControl.Padding = If(tutorial.ShowWithPaddings, DevExpress.XtraSpreadsheet.Demos.DemosInfo.DefaultPadding, System.Windows.Forms.Padding.Empty)
                Dim oldModule As DevExpress.DXperience.Demos.ModuleInfo = Nothing
                Dim oldTutorial As System.Windows.Forms.Control = Nothing
                If DevExpress.DXperience.Demos.ModulesInfo.Instance.CurrentModuleBase IsNot Nothing Then
                    If Equals(DevExpress.DXperience.Demos.ModulesInfo.Instance.CurrentModuleBase.Name, name) Then Return
                    oldModule = DevExpress.DXperience.Demos.ModulesInfo.Instance.CurrentModuleBase
                    oldTutorial = DevExpress.DXperience.Demos.ModulesInfo.Instance.CurrentModuleBase.TModule
                End If

                tutorial.Bounds = groupControl.DisplayRectangle
                DevExpress.DXperience.Demos.ModulesInfo.Instance.CurrentModuleBase = item
                tutorial.Visible = False
                groupControl.Controls.Add(tutorial)
                tutorial.Dock = System.Windows.Forms.DockStyle.Fill
                tutorial.RibbonMenuManager = manager
                tutorial.TutorialName = name
                tutorial.Visible = True
                item.WasShown = True
                Call DevExpress.XtraSpreadsheet.Demos.DemosInfo.AddToMru(item, tutorial)
                If oldTutorial IsNot Nothing Then
                    oldTutorial.Visible = False
                    groupControl.Controls.Remove(oldTutorial)
                End If

                Call DevExpress.XtraSpreadsheet.Demos.DemosInfo.PurgeMru()
            Finally
                System.Windows.Forms.Cursor.Current = currentCursor
            End Try

            Call DevExpress.DXperience.Demos.ModulesInfo.RaiseModuleChanged()
        End Sub

        Private Shared Sub AddToMru(ByVal [module] As DevExpress.DXperience.Demos.ModuleInfo, ByVal tutorial As System.Windows.Forms.Control)
            For i As Integer = 0 To DevExpress.XtraSpreadsheet.Demos.DemosInfo.mruList.Count - 1
                Dim item As DevExpress.XtraSpreadsheet.Demos.DemoMruItem = DevExpress.XtraSpreadsheet.Demos.DemosInfo.mruList(i)
                If Object.ReferenceEquals(item.[Module], [module]) Then
                    If i > 0 Then
                        Call DevExpress.XtraSpreadsheet.Demos.DemosInfo.mruList.RemoveAt(i)
                        Call DevExpress.XtraSpreadsheet.Demos.DemosInfo.mruList.Insert(0, item)
                    End If

                    Return
                End If
            Next

            Call DevExpress.XtraSpreadsheet.Demos.DemosInfo.mruList.Insert(0, New DevExpress.XtraSpreadsheet.Demos.DemoMruItem() With {.[Module] = [module], .TutorialControl = tutorial})
        End Sub

        Private Shared Sub PurgeMru()
            While DevExpress.XtraSpreadsheet.Demos.DemosInfo.mruList.Count > DevExpress.XtraSpreadsheet.Demos.DemosInfo.MaxMruCount
                Dim item As DevExpress.XtraSpreadsheet.Demos.DemoMruItem = DevExpress.XtraSpreadsheet.Demos.DemosInfo.mruList(DevExpress.XtraSpreadsheet.Demos.DemosInfo.mruList.Count - 1)
                item.TutorialControl.Dispose()
                item.[Module].ResetModule()
                Call DevExpress.XtraSpreadsheet.Demos.DemosInfo.mruList.RemoveAt(DevExpress.XtraSpreadsheet.Demos.DemosInfo.mruList.Count - 1)
            End While
        End Sub
    End Class
End Namespace

File: WinForms/CS/SpreadsheetMainDemo/DemosRegistration.cs
using System.Collections.Generic;
using System.Windows.Forms;
using DevExpress.DXperience.Demos;

namespace DevExpress.XtraSpreadsheet.Demos {
    public class DemosInfo : ModulesInfo {
        const int MaxMruCount = 3;
        static readonly Padding DefaultPadding = new Padding(8);
        static readonly List<DemoMruItem> mruList = new List<DemoMruItem>();
        public static void DoShowModule(string name, DevExpress.XtraEditors.GroupControl groupControl, SpreadsheetRibbonMenuManager manager) {
            ModuleInfo item = DemosInfo.GetItem(name);
            Cursor currentCursor = Cursor.Current;
            Cursor.Current = Cursors.WaitCursor;
            try {
                SpreadSheetTutorialControlBase tutorial = item.TModule as SpreadSheetTutorialControlBase;
                groupControl.Padding = tutorial.ShowWithPaddings ? DefaultPadding : Padding.Empty;

                ModuleInfo oldModule = null;
                Control oldTutorial = null;
                if(Instance.CurrentModuleBase != null) {
                    if(Instance.CurrentModuleBase.Name == name)
                        return;
                    oldModule = Instance.CurrentModuleBase;
                    oldTutorial = Instance.CurrentModuleBase.TModule;
                }

                tutorial.Bounds = groupControl.DisplayRectangle;
                Instance.CurrentModuleBase = item;
                tutorial.Visible = false;
                groupControl.Controls.Add(tutorial);
                tutorial.Dock = DockStyle.Fill;
                tutorial.RibbonMenuManager = manager;
                tutorial.TutorialName = name;
                tutorial.Visible = true;
                item.WasShown = true;
                AddToMru(item, tutorial);

                if(oldTutorial != null) {
                    oldTutorial.Visible = false;
                    groupControl.Controls.Remove(oldTutorial);
                }

                PurgeMru();
            }
            finally {
                Cursor.Current = currentCursor;
            }
            RaiseModuleChanged();
        }

        static void AddToMru(ModuleInfo module, Control tutorial) {
            for(int i = 0; i < mruList.Count; i++) {
                DemoMruItem item = mruList[i];
                if(object.ReferenceEquals(item.Module, module)) {
                    if(i > 0) {
                        mruList.RemoveAt(i);
                        mruList.Insert(0, item);
                    }
                    return;
                }
            }
            mruList.Insert(0, new DemoMruItem() { Module = module, TutorialControl = tutorial });
        }

        static void PurgeMru() {
            while(mruList.Count > MaxMruCount) {
                DemoMruItem item = mruList[mruList.Count - 1];
                item.TutorialControl.Dispose();
                item.Module.ResetModule();
                mruList.RemoveAt(mruList.Count - 1);
            }
        }
    }
}
